File: 2019/Jan/Bronze/3. Guess the Animal/guess.cc
/// AC

#include <bits/stdc++.h>

// #include <cstdio>
// #include <ios>
// #include <iostream>
// #include <string>
// #include <unordered_map>
// #include <vector>

using namespace std;

struct Bits {
  vector<int> data_;

  void set(size_t index) {
    if (index >= data_.size()) {
      data_.resize(index + 1);
    }
    data_[index] = 1;
  }
};

int count_both_set_bits(Bits const& set1, Bits const& set2) {
  int count = 0;
  for (int i = 0; i < min(set1.data_.size(), set2.data_.size()); i++) {
    if (set1.data_[i] && set2.data_[i]) {
      count++;
    }
  }
  return count;
}

int main() {
  //   ios::sync_with_stdio(false);
  //   input.tie(nullptr);

  ifstream input("guess.in");
  ofstream output("guess.out");

  int N, K;
  string animal, characteristic;

  input >> N;
  unordered_map<string, size_t> ch2index;
  size_t num_ch = 0;
  vector<Bits> animals;
  for (int i = 0; i < N; i++) {
    input >> animal >> K;
    Bits ch_bits;
    for (int j = 0; j < K; j++) {
      input >> characteristic;

      size_t index;
      auto it = ch2index.find(characteristic);
      if (it == ch2index.end()) {
        ch2index[characteristic] = index = num_ch;
        num_ch++;
      } else {
        index = it->second;
      }
      ch_bits.set(index);
    }
    animals.push_back(ch_bits);
  }

  int result = 0;
  for (int i = 0; i < animals.size(); i++) {
    for (int j = i + 1; j < animals.size(); j++) {
      int n = count_both_set_bits(animals[i], animals[j]);
      result = max(result, n);
    }
  }

  output << result + 1 << endl;
  return 0;
}

File: 2019/Jan/Bronze/3. Guess the Animal/guess.py
'''
AC
'''

read = open("guess.in", "r").readline
write = open("guess.out", "w").write


class Bits:
    def __init__(self) -> None:
        self.data = []

    def size(self) -> int:
        return len(self.data)

    def get(self, index) -> bool:
        return 1 == self.data[index]

    def set(self, index) -> None:
        if (index >= len(self.data)):
            for n in range(len(self.data), index + 1):
                self.data.append(0)
        self.data[index] = 1


def count_both_set_bits(bits1, bits2) -> int:
    count = 0
    for i in range(min(bits1.size(), bits2.size())):
        if bits1.get(i) and bits2.get(i):
            count += 1
    return count


N = int(read())
ch2index = {}
num_ch = 0
animals = []

for i in range(N):
    line = read()
    strs = line.split()
    animal = strs[0]
    K = int(strs[1])

    ch_bits = Bits()
    for ch in strs[2:]:
        if ch in ch2index:
            index = ch2index[ch]
        else:
            index = num_ch
            ch2index[ch] = index
            num_ch += 1
        ch_bits.set(index)
    animals.append(ch_bits)

result = 0
for i in range(len(animals)):
    for j in range(i+1, len(animals)):
        n = count_both_set_bits(animals[i], animals[j])
        result = max(result, n)

write(str(result + 1))
